tests/readers/test_batch_utils.py

```diff
@@ -0,0 +1,111 @@
+import numpy as np
+import pytest
+
+import tiledb
+from tiledb.ml.readers._batch_utils import estimate_row_bytes, get_num_batches
+
+
+@pytest.fixture
+def dense_uri(tmp_path):
+    uri = str(tmp_path / "dense")
+    schema = tiledb.ArraySchema(
+        sparse=False,
+        domain=tiledb.Domain(
+            tiledb.Dim(name="d0", domain=(0, 999), dtype=np.uint32),
+            tiledb.Dim(name="d1", domain=(1, 5), dtype=np.uint32),
+            tiledb.Dim(name="d2", domain=(1, 2), dtype=np.uint32),
+        ),
+        attrs=[
+            tiledb.Attr(name="af8", dtype=np.float64),
+            tiledb.Attr(name="af4", dtype=np.float32),
+            tiledb.Attr(name="au1", dtype=np.uint8),
+        ],
+    )
+    tiledb.Array.create(uri, schema)
+    with tiledb.open(uri, "w") as a:
+        size = a.schema.domain.size
+        a[:] = {
+            "af8": np.random.rand(size),
+            "af4": np.random.rand(size).astype(np.float32),
+            "au1": np.random.randint(128, size=size, dtype=np.uint8),
+        }
+    return uri
+
+
+@pytest.fixture
+def sparse_uri(tmp_path):
```

Comment:
Do we need separate fixtures for that or we could merge them depending on sparse flag? On the other hand offers modularity and informative tests.

Reply:
Unlike the rest reader tests, the `estimate_row_bytes` and `get_num_batches` tests (as well as the respective functions) depend heavilty on the array flavor (dense or sparse), it's not just a pass-through parameter to parametrize on. Merging them would require an "if sparse: ... else: ..." in the test logic, which defeats the point of merging them.